File: server/src/routes/play.ts
/**
 * Play Routes — interactive parent-child features for the baby-ai app.
 *
 * Endpoints defined in this file:
 *   POST /play/story       – SSE-streamed story generation (child-age–aware)
 *   POST /play/riddle      – AI-generated riddle (answer stored server-side)
 *   POST /play/riddle/guess – Verify a riddle answer or request a hint
 *   POST /play/baby-talk   – Interpret baby behaviour descriptions as "baby talk"
 *
 * All endpoints use Zod schemas for request-body validation (via validateBody
 * middleware). Riddle answers are held in an in-memory Map with a 5-minute TTL.
 *
 * @module playRouter
 */

import { Router } from 'express';
import { v4 as uuidv4 } from 'uuid';
import { z } from 'zod';
import { validateBody } from '../middleware/validateBody.js';
import { chat, chatStream } from '../services/deepseek.js';
import { initSSE, sendToken, sendDone, sendError } from '../utils/sse.js';
import { STORY_PROMPT, RIDDLE_PROMPT, BABY_TALK_PROMPT } from '../services/playPrompts.js';
import logger from '../utils/logger.js';

export const playRouter = Router();

/* ---- Zod Schemas ---- */

/** Schema for POST /play/story — validates a story-generation request. */
const StoryRequestSchema = z.object({
  /** Child's age in years (0–12). */
  childAge: z.number().int().min(0).max(12),
  /** Optional interest keyword (e.g. "dinosaurs"); max 100 chars. */
  interest: z.string().max(100).optional(),
  /** Optional story genre: bedtime, adventure, or educational. */
  storyType: z.enum(['bedtime', 'adventure', 'educational']).optional(),
});

/** Schema for POST /play/riddle — validates a riddle-generation request. */
const RiddleRequestSchema = z.object({
  /** Child's age in years (0–12). */
  childAge: z.number().int().min(0).max(12),
  /** Optional difficulty level; defaults to AI-determined when omitted. */
  difficulty: z.enum(['easy', 'medium', 'hard']).optional(),
});

/** Schema for POST /play/riddle/guess — validates a guess or hint request. */
const RiddleGuessSchema = z.object({
  /** The riddle ID returned by POST /play/riddle. */
  riddleId: z.string(),
  /** The child's answer string (omit when requesting a hint). */
  guess: z.string().optional(),
  /** Set to true to request the next hint instead of submitting a guess. */
  hint: z.boolean().optional(),
});

/** Schema for POST /play/baby-talk — validates a baby-talk interpretation request. */
const BabyTalkRequestSchema = z.object({
  /** Description of what the baby is doing / how they are behaving (2–500 chars). */
  description: z.string().min(2).max(500),
  /** Optional baby age in months (0–36); improves contextual interpretation. */
  babyAge: z.number().int().min(0).max(36).optional(),
});

/**
 * In-memory store for active riddles.
 *
 * Each entry maps a riddleId to its answer, available hints, and creation
 * timestamp. Entries expire after 5 minutes (300 000 ms) — a background
 * interval sweeps the map every 60 seconds.
 *
 * Edge case: if the server restarts all riddles are lost; clients receive a
 * 404 with code RIDDLE_EXPIRED and should ask for a new riddle.
 */
const riddleStore = new Map<string, { answer: string; hints: string[]; createdAt: number }>();

// Periodically purge expired riddles (TTL = 5 minutes).
setInterval(() => {
  const now = Date.now();
  for (const [key, val] of riddleStore) {
    if (now - val.createdAt > 300000) riddleStore.delete(key);
  }
}, 60000);

/* ---- Routes ---- */

/**
 * POST /play/story — generate a child-friendly story via SSE stream.
 *
 * Builds a natural-language prompt from the child's age, optional interest,
 * and optional story type, then streams tokens back to the client over
 * Server-Sent Events (SSE).
 *
 * @param {Request}  req  – Express request; body validated by StoryRequestSchema.
 * @param {Response} res  – Express response; SSE stream with real-time token
 *   delivery. The stream ends with a `sendDone` frame carrying the messageId
 *   and a usage token count, or a `sendError` frame on failure.
 *
 * Method:  POST
 * Path:    /play/story
 *
 * Error scenarios:
 *   - DeepSeek API failure → SSE error frame with code STORY_ERROR.
 */
playRouter.post('/story', validateBody(StoryRequestSchema), async (req, res) => {
  // Extract validated fields from the request body.
  const { childAge, interest, storyType } = req.body;

  // Compose the user-facing message that seeds the story generation.
  let userMessage = `请为${childAge}岁的小朋友创作一个故事。`;
  if (storyType) {
    // Map the enum values to Chinese genre labels.
    const typeMap: Record<string, string> = { bedtime: '睡前故事', adventure: '冒险故事', educational: '教育故事' };
    userMessage += `故事类型：${typeMap[storyType]}。`;
  }
  if (interest) {
    userMessage += `小朋友喜欢：${interest}。`;
  }

  // Initialize the SSE connection — sets headers and keeps the response open.
  initSSE(res);
  const messageId = uuidv4();

  try {
    // Stream tokens from DeepSeek one chunk at a time.
    for await (const chunk of chatStream([
      { role: 'system', content: STORY_PROMPT },
      { role: 'user', content: userMessage },
    ])) {
      sendToken(res, chunk);
    }
    // Signal successful completion with a final SSE frame.
    sendDone(res, messageId, 0);
    res.end();
  } catch (err: any) {
    logger.error('故事生成失败:', err.message);
    // Send a structured SSE error so the client can display a friendly message.
    sendError(res, 'STORY_ERROR', '生成故事时出现错误');
    res.end();
  }
});

/**
 * POST /play/riddle — generate a new riddle for a child.
 *
 * Calls DeepSeek with a system prompt tuned for child-appropriate riddles.
 * The returned JSON is parsed to extract the riddle text, answer, and hints.
 * The answer is stored server-side (in-memory) and never sent to the client.
 *
 * @param {Request}  req  – Express request; body validated by RiddleRequestSchema.
 * @param {Response} res  – Express response; JSON object containing:
 *   - `riddleId`   – unique ID the client uses for subsequent guess requests
 *   - `riddle`     – the riddle question text
 *   - `category`   – topic category (e.g. "动物", "生活")
 *   - `difficulty` – difficulty level
 * @param {NextFunction} next – Express next function; passed the error if
 *   DeepSeek fails.
 *
 * Method:  POST
 * Path:    /play/riddle
 *
 * Error scenarios:
 *   - AI does not return valid JSON → fallback to a default riddle shape so
 *     the game can continue (answer is set to a placeholder; hints are minimal).
 *   - DeepSeek API failure → forwarded to the global error handler via next(err).
 */
playRouter.post('/riddle', validateBody(RiddleRequestSchema), async (req, res, next) => {
  // Extract validated fields.
  const { childAge, difficulty } = req.body;

  // Build the AI prompt with age and optional difficulty.
  let userMessage = `请为${childAge}岁的小朋友设计一个谜语。`;
  if (difficulty) userMessage += `难度：${difficulty}。`;

  try {
    // Non-streaming call — we need the full JSON response at once.
    const { content } = await chat([
      { role: 'system', content: RIDDLE_PROMPT },
      { role: 'user', content: userMessage },
    ], { temperature: 0.8 });

    /**
     * Parse the AI response as JSON.
     *
     * Edge case: the model may wrap the JSON in markdown code fences (```json … ```).
     * We attempt a regex extraction first; if that fails we try a direct parse.
     * If both fail we fall back to a safe default so the game does not break.
     */
    let riddleData: any;
    try {
      // Try to extract the first JSON object from the response text.
      const jsonMatch = content.match(/\{[\s\S]*\}/);
      riddleData = jsonMatch ? JSON.parse(jsonMatch[0]) : JSON.parse(content);
    } catch {
      // Fallback: treat the entire response as the riddle text.
      riddleData = {
        riddle: content.trim(),
        category: '生活',
        difficulty: 'easy',
        answer: '（答案生成失败）',
        hints: ['再想想吧~'],
      };
    }

    // Persist answer + hints server-side so the client cannot peek.
    const riddleId = `rid_${uuidv4().slice(0, 8)}`;
    riddleStore.set(riddleId, {
      answer: riddleData.answer || '',
      hints: riddleData.hints || [],
      createdAt: Date.now(),
    });

    // Return only the public-facing fields (answer is omitted).
    res.json({
      riddleId,
      riddle: riddleData.riddle,
      category: riddleData.category,
      difficulty: riddleData.difficulty,
    });
  } catch (err) {
    // Forward to Express global error handler.
    next(err);
  }
});

/**
 * POST /play/riddle/guess — submit a guess or request a hint for an active
 * riddle.
 *
 * The endpoint supports two modes controlled by the `hint` flag in the body:
 *   - hint=true  → returns the next available hint for the riddle.
 *   - hint=false → evaluates the submitted guess against the stored answer.
 *
 * @param {Request}  req  – Express request; body validated by RiddleGuessSchema.
 * @param {Response} res  – Express response.
 *
 *   Guess mode (hint omitted or false):
 *     - Correct   → `{ correct: true,  answer, encouragement }`  (riddle deleted)
 *     - Incorrect → `{ correct: false, encouragement }`
 *
 *   Hint mode (hint=true):
 *     - `{ hint, correct: false }`
 *
 *   Error mode:
 *     - 404 `{ error, code: 'RIDDLE_EXPIRED' }` when the riddle is unknown/expired.
 *     - 400 `{ error, code: 'NO_GUESS' }` when a guess is expected but missing/empty.
 *
 * Method:  POST
 * Path:    /play/riddle/guess
 *
 * Edge cases:
 *   - Concurrent hint requests consume hints one at a time (shift).
 *   - Answer matching uses fuzzy logic: exact match, substring containment
 *     (guess includes answer, or answer includes guess) to be lenient with
 *     children's input.
 *   - On a correct guess the riddle is immediately deleted to prevent replay.
 */
playRouter.post('/riddle/guess', validateBody(RiddleGuessSchema), async (req, res) => {
  const { riddleId, guess, hint } = req.body;

  // Look up the riddle in the in-memory store.
  const stored = riddleStore.get(riddleId);

  // If the riddle has expired or never existed, tell the client to request a new one.
  if (!stored) {
    res.status(404).json({ error: '谜语已过期，请重新出题', code: 'RIDDLE_EXPIRED' });
    return;
  }

  // --- Hint mode ---
  if (hint) {
    const nextHintIdx = 0; // 简单实现：每次给第一个还没给的提示
    // Consume the first available hint (destructive — each hint is given only once).
    const hintText = stored.hints.length > 0 ? stored.hints[0] : '没有更多提示啦~';
    stored.hints.shift(); // Remove the hint so the next request gets a different one.
    res.json({ hint: hintText, correct: false });
    return;
  }

  // --- Guess mode ---

  // Guard against empty / whitespace-only input.
  if (!guess || !guess.trim()) {
    res.status(400).json({ error: '请输入你的答案', code: 'NO_GUESS' });
    return;
  }

  /**
   * Fuzzy match: accept an exact match, or if either string is a substring of
   * the other. This makes the experience friendlier for young children who may
   * type partial answers.
   */
  const isCorrect = guess.trim() === stored.answer ||
    guess.trim().includes(stored.answer) ||
    stored.answer.includes(guess.trim());

  if (isCorrect) {
    // Delete the riddle immediately — one correct guess per riddle.
    riddleStore.delete(riddleId);
    // Randomly select a positive encouragement message.
    const encouragements = [
      '太棒了！你真聪明！🌟',
      '答对啦！你好厉害！🎉',
      '完全正确！看来你是个猜谜高手！🏆',
    ];
    res.json({
      correct: true,
      answer: stored.answer,
      encouragement: encouragements[Math.floor(Math.random() * encouragements.length)],
    });
  } else {
    res.json({
      correct: false,
      encouragement: '不对哦，再想想！💪',
    });
  }
});

/**
 * POST /play/baby-talk — interpret baby behaviour as "what the baby would say."
 *
 * Takes a free-text description of a baby's actions / sounds and optionally the
 * baby's age in months, then returns a playful "translation" generated by the
 * AI. Useful for parents who want a fun, light-hearted interpretation.
 *
 * @param {Request}  req  – Express request; body validated by BabyTalkRequestSchema.
 * @param {Response} res  – Express response; JSON `{ interpretation: string }`.
 * @param {NextFunction} next – Express next function; forwards errors to the
 *   global error handler.
 *
 * Method:  POST
 * Path:    /play/baby-talk
 *
 * Error scenarios:
 *   - DeepSeek API failure → forwarded to global error handler via next(err).
 *   - Empty / overly short description → rejected by Zod validation (min 2 chars).
 */
playRouter.post('/baby-talk', validateBody(BabyTalkRequestSchema), async (req, res, next) => {
  // Extract validated fields.
  const { description, babyAge } = req.body;

  // Build the prompt. Baby age is optional and improves context.
  let userMessage = `宝宝表现描述：${description}`;
  if (babyAge !== undefined) {
    userMessage += `\n宝宝年龄：${babyAge}个月`;
  }

  try {
    // Non-streaming call with higher temperature for creative outputs.
    const { content } = await chat([
      { role: 'system', content: BABY_TALK_PROMPT },
      { role: 'user', content: userMessage },
    ], { temperature: 0.9, maxTokens: 500 });

    // Return the AI's interpretation as a plain JSON field.
    res.json({ interpretation: content });
  } catch (err) {
    // Forward to Express global error handler.
    next(err);
  }
});

File: python-server/app/api/play.py
"""
趣味互动路由（Play Routes）

从原 Express 后端 `server/src/routes/play.ts` 迁移而来，接口契约保持不变：
  - POST /api/play/story        —— SSE 流式生成儿童故事
  - POST /api/play/riddle       —— AI 生成谜语（答案存服务端，不返回给前端）
  - POST /api/play/riddle/guess —— 校验谜底 / 请求提示
  - POST /api/play/baby-talk    —— "婴语翻译"趣味解读

设计要点（面试常问）：
  1. 谜语答案用进程内 TTL 缓存（TTLCache，5 分钟过期）存服务端，前端只能拿到
     riddleId，无法"作弊"看到答案；猜对立即 delete 防重放。
  2. 谜语生成要求 LLM 输出严格 JSON，解析失败时降级到兜底谜语，保证游戏不崩。
  3. 答案匹配用模糊逻辑（完全相等 / 子串包含），宽容儿童的近似输入。
"""

import json
import random
import re

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse, StreamingResponse

from app.core.cache import TTLCache
from app.models.chat import ChatMessage, ChatOptions
from app.models.play import (
    BabyTalkRequest,
    RiddleGuessRequest,
    RiddleRequest,
    StoryRequest,
)
from app.services.llm import call_deepseek, generate_stream_with_interrupt_and_fallback
from app.services.play_prompts import BABY_TALK_PROMPT, RIDDLE_PROMPT, STORY_PROMPT
from app.utils.logger import logger

router = APIRouter(prefix="/api/play", tags=["play"])

# 谜语答案存储：进程内 TTL 缓存，5 分钟过期（对齐原 TS 后端的 Map + setInterval 清理）
# value 结构：{"answer": str, "hints": list[str]}
riddle_store = TTLCache(maxsize=512, ttl=300)

_STORY_TYPE_MAP = {"bedtime": "睡前故事", "adventure": "冒险故事", "educational": "教育故事"}


# ===============================================================
# 儿童故事生成（SSE）
# ===============================================================

@router.post("/story")
async def story(data: StoryRequest, request: Request):
    """流式生成儿童故事，token 包装成 SSE 事件（对齐 chat 流式格式）"""
    user_message = f"请为{data.childAge}岁的小朋友创作一个故事。"
    if data.storyType:
        user_message += f"故事类型：{_STORY_TYPE_MAP[data.storyType]}。"
    if data.interest:
        user_message += f"小朋友喜欢：{data.interest}。"

    messages = [
        ChatMessage(role="system", content=STORY_PROMPT),
        ChatMessage(role="user", content=user_message),
    ]

    async def gen():
        async for chunk in generate_stream_with_interrupt_and_fallback(messages, request):
            yield f"event: token\ndata: {json.dumps({'text': chunk}, ensure_ascii=False)}\n\n"
        yield "event: done\ndata: {}\n\n"

    return StreamingResponse(gen(), media_type="text/event-stream")


# ===============================================================
# 谜语生成（非流式，答案存服务端）
# ===============================================================

@router.post("/riddle")
def riddle(data: RiddleRequest):
    """生成谜语；答案 + 提示存服务端，仅返回公开字段"""
    user_message = f"请为{data.childAge}岁的小朋友设计一个谜语。"
    if data.difficulty:
        user_message += f"难度：{data.difficulty}。"

    content = call_deepseek(
        [
            ChatMessage(role="system", content=RIDDLE_PROMPT),
            ChatMessage(role="user", content=user_message),
        ],
        ChatOptions(temperature=0.8),
    )

    # 解析 LLM 返回的 JSON：先正则提取（兼容 ```json 代码块包裹），再直接 parse，
    # 都失败则降级到兜底谜语，保证游戏不中断。
    try:
        match = re.search(r"\{[\s\S]*\}", content)
        riddle_data = json.loads(match.group(0)) if match else json.loads(content)
    except (json.JSONDecodeError, AttributeError):
        logger.warning("[play] 谜语 JSON 解析失败，走兜底谜语")
        riddle_data = {
            "riddle": content.strip() or "我圆圆的、红红的，长在树上，咬一口甜甜的，是什么呀？",
            "category": "生活",
            "difficulty": "easy",
            "answer": "苹果",
            "hints": ["它是一种水果哦~", "它的颜色很喜庆", "圣诞节常用来装饰"],
        }

    riddle_id = f"rid_{random.randbytes(4).hex()}"
    riddle_store.set(
        riddle_id,
        {"answer": (riddle_data.get("answer") or "").strip(), "hints": riddle_data.get("hints") or []},
    )

    return {
        "riddleId": riddle_id,
        "riddle": riddle_data.get("riddle", ""),
        "category": riddle_data.get("category", "生活"),
        "difficulty": riddle_data.get("difficulty", "easy"),
    }


# ===============================================================
# 猜谜 / 求提示
# ===============================================================

@router.post("/riddle/guess")
def guess(data: RiddleGuessRequest):
    """提交猜测或请求下一个提示"""
    stored = riddle_store.get(data.riddleId)
    if stored is None:
        return JSONResponse(
            status_code=404,
            content={"error": "谜语已过期，请重新出题", "code": "RIDDLE_EXPIRED"},
        )

    # 提示模式：每次给第一个还没给的提示（shift，消费即删）
    if data.hint:
        hint_text = stored["hints"][0] if stored["hints"] else "没有更多提示啦~"
        if stored["hints"]:
            stored["hints"].pop(0)
        return {"hint": hint_text, "correct": False}

    # 猜谜模式：空输入校验
    if not data.guess or not data.guess.strip():
        return JSONResponse(
            status_code=400,
            content={"error": "请输入你的答案", "code": "NO_GUESS"},
        )

    answer = stored["answer"]
    guess_text = data.guess.strip()
    # 模糊匹配：完全相等 / 互相子串包含（宽容儿童近似输入）
    is_correct = guess_text == answer or guess_text in answer or answer in guess_text

    if is_correct:
        riddle_store.delete(data.riddleId)  # 答对即删，防重放
        encouragement = random.choice([
            "太棒了！你真聪明！🌟",
            "答对啦！你好厉害！🎉",
            "完全正确！看来你是个猜谜高手！🏆",
        ])
        return {"correct": True, "answer": answer, "encouragement": encouragement}

    return {"correct": False, "encouragement": "不对哦，再想想！💪"}


# ===============================================================
# 婴语翻译（非流式）
# ===============================================================

@router.post("/baby-talk")
def baby_talk(data: BabyTalkRequest):
    """趣味解读宝宝行为，返回"宝宝内心 OS"翻译"""
    user_message = f"宝宝表现描述：{data.description}"
    if data.babyAge is not None:
        user_message += f"\n宝宝年龄：{data.babyAge}个月"

    content = call_deepseek(
        [
            ChatMessage(role="system", content=BABY_TALK_PROMPT),
            ChatMessage(role="user", content=user_message),
        ],
        ChatOptions(temperature=0.9, maxTokens=500),
    )
    return {"interpretation": content}
